modules/3d/include/opencv2/3d/depth.hpp

```diff
@@ -0,0 +1,502 @@
+
+    CV_WRAP virtual void setPyramidLevels(size_t _nLevels) = 0;
+    CV_WRAP virtual size_t getPyramidLevels(int pyrType) = 0;
+
+    CV_WRAP virtual void setPyramidAt(InputArray  _pyrImage, int pyrType, size_t level) = 0;
+    CV_WRAP virtual void getPyramidAt(OutputArray _pyrImage, int pyrType, size_t level) = 0;
+
+    CV_PROP int ID;
+};
+
+
+/** Base class for computation of odometry.
+ */
+class CV_EXPORTS_W Odometry
+{
+public:
+
+    /** A class of transformation*/
+    enum
+    {
+      ROTATION = 1, TRANSLATION = 2, RIGID_BODY_MOTION = 4
+    };
+
+    virtual ~Odometry() { }
+
+    /** Create a new Odometry object based on its name. Currently supported names are:
+    * "RgbdOdometry", "ICPOdometry", "RgbdICPOdometry", "FastICPOdometry".
+    * @param odometryType algorithm's name
+    */
+    CV_WRAP static Ptr<Odometry> createFromName(const String& odometryType);
+
+    /** Method to compute a transformation from the source frame to the destination one.
+     * Some odometry algorithms do not used some data of frames (eg. ICP does not use images).
+     * In such case corresponding arguments can be set as empty Mat.
+     * The method returns true if all internal computations were possible (e.g. there were enough correspondences,
+     * system of equations has a solution, etc) and resulting transformation satisfies some test if it's provided
+     * by the Odometry inheritor implementation (e.g. thresholds for maximum translation and rotation).
+     * @param srcImage Image data of the source frame (CV_8UC1)
+     * @param srcDepth Depth data of the source frame (CV_32FC1, in meters)
+     * @param srcMask Mask that sets which pixels have to be used from the source frame (CV_8UC1)
+     * @param dstImage Image data of the destination frame (CV_8UC1)
+     * @param dstDepth Depth data of the destination frame (CV_32FC1, in meters)
+     * @param dstMask Mask that sets which pixels have to be used from the destination frame (CV_8UC1)
+     * @param Rt Resulting transformation from the source frame to the destination one (rigid body motion):
+     dst_p = Rt * src_p, where dst_p is a homogeneous point in the destination frame and src_p is
+     homogeneous point in the source frame,
+     Rt is 4x4 matrix of CV_64FC1 type.
+     * @param initRt Initial transformation from the source frame to the destination one (optional)
+     */
+    CV_WRAP virtual bool compute(InputArray srcImage, InputArray srcDepth, InputArray srcMask, InputArray dstImage, InputArray dstDepth,
+                                 InputArray dstMask, OutputArray Rt, const Mat& initRt = Mat()) const = 0;
+
+    /** One more method to compute a transformation from the source frame to the destination one.
+     * It is designed to save on computing the frame data (image pyramids, normals, etc.).
+     */
+    CV_WRAP_AS(compute2) virtual bool compute(Ptr<OdometryFrame> srcFrame, Ptr<OdometryFrame> dstFrame, OutputArray Rt, const Mat& initRt = Mat()) const = 0;
```

Comment:
> Ptr<OdometryFrame>

"const reference" for parameters

Reply:
1. The reference is not an option: `OdometryFrame` is a pure abstract class with two children: for Mats and for UMats inside. Not sure the reference can be used here without making one more layer of pImpl for `OdometryFrame` which seem to tangle the code much more then it's right now.
2. Using const is not an option: `compute()` calls `prepareFrameCache()` inside and saves that cache in `OdometryFrame` fields for further computations. Of course, making its fields mutable should work but it does not make sense since all the fields are used for caching, so there would be fake constness.

Reply:
`const Ptr<OdometryFrame>& srcFrame`
`const Ptr<OdometryFrame>& dstFrame`
`InputArray initRt = noArray()`

Reply:
Thanks, now I understand!
Fixed + all Mats in `depth.hpp` replaced by `InputArray`s